pageserver/src/tenant/ephemeral_file.rs

```diff
@@ -51,57 +65,186 @@ impl EphemeralFile {
+        let size = usize::try_from(self.len()).unwrap();
+        let vec = Vec::with_capacity(size);
+
+        // read from disk what we've already flushed
+        let file_size_tracker = self.buffered_writer.as_inner();
+        let flushed_offset = usize::try_from(file_size_tracker.bytes_written()).unwrap();
+        let flushed_range = 0..flushed_offset;
+        let file: &VirtualFile = file_size_tracker.as_inner();
+        let mut vec = file
+            .read_exact_at(
+                vec.slice(0..(flushed_range.end - flushed_range.start)),
+                u64::try_from(flushed_range.start).unwrap(),
+                ctx,
+            )
+            .await?
+            .into_inner();
+
+        // copy from in-memory buffer what we haven't flushed yet but would return when accessed via read_blk
+        let buffer = self.buffered_writer.inspect_buffer();
+        let buffered = &buffer[0..buffer.pending()];
+        vec.extend_from_slice(buffered);
+        assert_eq!(vec.len(), size);
+        Ok(vec)
     }
 
-    pub(crate) async fn read_blk(
+    /// Fill dst will dst.bytes_total() bytes from the bytes written to the buffered writer from offset `start` and later.
+    /// If `dst` is larger than the available bytes, the read will be short.
+    /// The read will never be short for other reasons.
+    /// The number of bytes read into `dst` is returned as part of the result tuple.
+    /// No guarantees are made about the remaining bytes in `dst`, i.e., assume their contents are random.
+    pub(crate) async fn read_at_to_end<B: IoBufMut + Send>(
         &self,
-        blknum: u32,
+        start: u32,
+        dst: Slice<B>,
         ctx: &RequestContext,
-    ) -> Result<BlockLease, io::Error> {
-        self.rw.read_blk(blknum, ctx).await
+    ) -> std::io::Result<(Slice<B>, usize)> {
+        let file_size_tracking_writer = self.buffered_writer.as_inner();
+        let flushed_offset = u32::try_from(file_size_tracking_writer.bytes_written())
+            .expect("we don't allow writing more than u32::MAX bytes");
+
+        let buffer = self.buffered_writer.inspect_buffer();
+        let buffered = &buffer[0..buffer.pending()];
+
+        let dst_cap = u32::try_from(dst.bytes_total())
+            .with_context(|| {
+                format!(
+                    "read_aligned: dst.bytes_total() is too large: {}",
+                    dst.len()
+                )
+            })
+            .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+        let end = {
+            let mut end = start
+                .checked_add(dst_cap)
+                .with_context(|| {
+                    format!("read_aligned: offset + dst.bytes_total() is too large: {start} + {dst_cap}",)
+                })
+                .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;
+            if end > self.bytes_written {
+                end = self.bytes_written;
+            }
+            end
```

Comment:
If the end of the ephemeral file is not aligned, doesn't this result in misaligned DIO read? Can we assume that all files are aligned? Probably not.

This really is a more general question of how do we handle pre-existing misaligned files with DIO.

Reply:
ext4 allows O_DIRECT reads from the end of the file, I checked that manually.

For `EphemeralFile` though, we don't actually need to worry because TAIL_SZ is a multiple of DIO_CHUNK_SIZE, so, the part of read_at_to_end that reads from the file will always send a read to the file that is a DIO_CHUNK_SIZE multiple.